src/bin/cargo/commands/remove.rs

```diff
@@ -131,39 +135,91 @@ fn gc_workspace(workspace: &Workspace<'_>, dependencies: &[String]) -> CargoResu
 
-/// Remove a dependency from a workspace manifest.
-fn remove_workspace_dep(dep: &str, ws_manifest: &mut toml_edit::Document) {
-    if let Some(toml_edit::Item::Table(table)) = ws_manifest
+    // clean up workspace.dependencies
+    if let Some(toml_edit::Item::Table(deps_table)) = manifest
         .get_mut("workspace")
         .and_then(|t| t.get_mut("dependencies"))
     {
+        deps_table.set_implicit(true);
+        for (key, item) in deps_table.iter_mut() {
+            if !dependencies.iter().any(|d| {
+                d.toml_key() == key.get() && matches!(d.source(), Some(Source::Workspace(_)))
+            }) {
+                *item = toml_edit::Item::None;
+            }
+        }
+    }
+
+    // clean up the profile section
+    if let Some(toml_edit::Item::Table(profile_section_table)) = manifest.get_mut("profile") {
+        profile_section_table.set_implicit(true);
+        for (_, item) in profile_section_table.iter_mut() {
+            if let toml_edit::Item::Table(profile_table) = item {
+                profile_table.set_implicit(true);
+                if let Some(toml_edit::Item::Table(package_table)) =
+                    profile_table.get_mut("package")
+                {
+                    package_table.set_implicit(true);
+                    for (key, item) in package_table.iter_mut() {
+                        if let Ok(spec) = PackageIdSpec::parse(key.get()) {
+                            if !dependencies.iter().any(|d| d.name == spec.name().as_str()) {
+                                *item = toml_edit::Item::None;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // clean up patch section
+    if let Some(toml_edit::Item::Table(patch_section_table)) = manifest.get_mut("patch") {
+        patch_section_table.set_implicit(true);
+        for (_, item) in patch_section_table.iter_mut() {
+            if let toml_edit::Item::Table(patch_table) = item {
+                patch_table.set_implicit(true);
+                for (key, item) in patch_table.iter_mut() {
+                    if !dependencies.iter().any(|d| d.name == key.get()) {
+                        *item = toml_edit::Item::None;
+                    }
+                }
+            }
+        }
+    }
+
+    // clean up replace section
+    if let Some(toml_edit::Item::Table(table)) = manifest.get_mut("replace") {
         table.set_implicit(true);
-        table.remove(dep);
+        for (key, item) in table.iter_mut() {
+            if let Ok(spec) = PackageIdSpec::parse(key.get()) {
+                if !dependencies.iter().any(|d| {
+                    d.name == spec.name().as_str()
+                        && d.version()
+                            .and_then(|v| semver::VersionReq::parse(v).ok())
+                            .and_then(|vq| spec.version().map(|v| vq.matches(v)))
+                            .unwrap_or(true)
```

Comment:
Maybe we should break this out into a function, both to make this code easier to use and for reuse with `profile`